### ISSUE TITLE
Breaking backwards compatibility

### ISSUE THREAD

**Issue by iainbeeston** [ruby-json-schema/json-schema#246]
There are many features in the gem that are no longer "desirable" and could be removed, to reduce complexity and make the project more maintainable. However they might break backwards compatibility.

For example:

* Stringifying input (vs expecting the user to provide a stringified hash)
* Auto-detecting whether an input string is a url (before downloading and parsing) or a JSON text (vs expecting the user to download and parse, and only accepting a valid JSON hash)
* Setting global options on the JSON::Schema class itself (vs. passing parameters in as options to validate, at runtime)
* Generating our own UUIDs (vs using SecureRandom)

I'm sure there are many others.

Lately I've been wondering whether supporting optional or infrequently used features (and functionality that can be done for us by the Ruby core lib or another gem) is a good thing, and whether the library would be easier to maintain without them. And if that is the case, should version 3 aim to strip out things we know are "legacy"? Another option might be for JSON schema to continue backwards compatibility, but be forked to produce a new gem which would have no need for backwards compatibility, and could have large parts of the code stripped out and be a more versatile base for the future. Or is the aim of the gem to preserve backwards compatibility for the sake of existing users?

Ideally I would like to see the gem doing as little as possible, sticking to the core requirement of validating JSON, and expecting the caller to write code for anything else that they might need, on a case-by-case basis (or by writing a separate gem for that).

I'd love to hear opinions on how we should handle optional or legacy features, especially from the core team.

**Comment by hoxworth** [comment]
Honestly, I agree and would love to see the move towards trying to be "just" validation moving forward. About the time I started losing the ability to spend a lot of time on this project was when I started just accepting all manner of feature requests, and the project became fairly bloated as a result. I'd prefer to keep this library be mostly just a validator that follow the spec as closely as possible with the added benefit of being able to define custom extensions. 

Everything else is cruft.

**Comment by iainbeeston** [comment]
Ah yes, being extendible would be a good idea. Maybe we could even refactor
some of the existing non-core features (inserting defaults?) out to
extension gems

**Comment by iainbeeston** [comment]
I've added a wish list of things that would be good to have for the 3.0 release in #247. It'd be great to have some editing/feedback from other maintainers on that.

**Comment by RST-J** [comment]
What is about the ``:list`` option? I think, this can also be easily done by the user and would simplify the code again.

**Comment by iainbeeston** [comment]
Yes, I think that would also be good to remove (can you add it to #247? So
that we don't forget)
